File: 450-delete-node-in-a-bst/450-delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
   TreeNode findMin(TreeNode temp) {
        while (temp.left != null) {
            temp = temp.left;
        }
        return temp;
    }

    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) {
            return null;
        } else if (root.val > key) {
            root.left = deleteNode(root.left, key);
        } else if (root.val < key) {
            root.right = deleteNode(root.right, key);
        } else {
            if (root.left == null && root.right == null) {
                root = null;
            } else if (root.left == null) {
                root = root.right;
            } else if (root.right == null) {
                root = root.left;
            } else {
                TreeNode min = findMin(root.right);
                root.val = min.val;
                root.right = deleteNode(root.right, min.val);
            }
        }
        return root;
    }
}

File: 450-delete-node-in-a-bst/450-delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode *findMinRight(TreeNode *root) {
        while(root->left != NULL) {
            root = root->left;
        }
        return root;
    }
    TreeNode* deleteNode(TreeNode* root, int key) {
        
        if(root == NULL) {
            return root;
        }
        
        if(root->val == key) {
            if(root->left == NULL) {
                TreeNode *current = root;
                root =  root->right;
                delete current;
            }
            else if(root->right == NULL){
                TreeNode *current = root;
                root =  root->left;
                delete current;
            }
            else {
                TreeNode *rightMin = findMinRight(root->right);
                swap(root->val,rightMin->val);
                root->right = deleteNode(root->right,key);
            }
        }
        else {
            if(key < root->val) {
                root->left  = deleteNode(root->left,key);
            }
            else {
                root->right = deleteNode(root->right,key);
            }
        }
        return root;
    }
};